src/devices/cpu/pic16x8x/pic16x8x.cpp

```diff
@@ -0,0 +1,1443 @@
+// | GIE | EEIE | T0IE | INTE | RBIE | T0IF | INTF | RBIF |
+
+#define GIE_FLAG   0x80
+#define EEIE_FLAG  0x40
+#define T0IE_FLAG  0x20
+#define INTE_FLAG  0x10
+#define RBIE_FLAG  0x08
+#define T0IF_FLAG  0x04
+#define INTF_FLAG  0x02
+#define RBIF_FLAG  0x01
+
+#define GIE   (m_INTCON & GIE_FLAG)
+#define EEIE  (m_INTCON & EEIE_FLAG)
+#define T0IE  (m_INTCON & T0IE_FLAG)
+#define INTE  (m_INTCON & INTE_FLAG)
+#define RBIE  (m_INTCON & RBIE_FLAG)
+#define T0IF  (m_INTCON & T0IF_FLAG)
+#define INTF  (m_INTCON & INTF_FLAG)
+#define RBIF  (m_INTCON & RBIF_FLAG)
+
+// ******** The following is the EECON1 Flag register definition. ********
+// | 7 | 6 | 5 |   4  |   3   |   2  |  1 |  0 |
+// |   |   |   | EEIF | WRERR | WREN | WR | RD |
+
+#define EEIF_FLAG  0x10
+#define WRERR_FLAG 0x08
+#define WREN_FLAG  0x04
+#define EEWR_FLAG  0x02
+#define EERD_FLAG  0x01
+
+#define EEIF  (m_EECON1 & EEIF_FLAG)
+#define WRERR (m_EECON1 & WRERR_FLAG)
+#define WREN  (m_EECON1 & WREN_FLAG)
+#define EEWR  (m_EECON1 & EEWR_FLAG)
+#define EERD  (m_EECON1 & EERD_FLAG)
+
+/************************************************************************
+ *  Fixed Vectors
+ ************************************************************************/
+
+#define RESET_VECTOR 0x00
+#define INT_VECTOR 0x04
+
+/************************************************************************
+ *  Shortcuts
+ ************************************************************************/
+
+#define RISING_EDGE_RB0   (( (int)(RB0_in - m_old_RB0) > 0) ? 1 : 0)
+#define FALLING_EDGE_RB0  (( (int)(RB0_in - m_old_RB0) < 0) ? 1 : 0)
+
+//#define S_RB0_IN          (m_read_port[PORTB](PORTB, 0x01))
+
+#define M_RDEEPROM(A)     m_eeread(A)
+#define M_WREEPROM(A,V)   m_eewrite(A, V)
+
+#define CLR(flagreg, flag)  (flagreg &= u8(~flag))
+#define SET(flagreg, flag)  (flagreg |= (flag))
+
+#define ADDR  ((m_opcode.b.l & 0x7f) | (RP0 << 2))
+#define BITPOS  ((m_opcode.w >> 7) & 7)
+
+
+void pic16x8x_device::calc_zero_flag()
+{
+	if (m_ALU == 0)
+		SET(m_STATUS, Z_FLAG);
+	else
+		CLR(m_STATUS, Z_FLAG);
+}
+
+void pic16x8x_device::calc_add_flags(u8 augend)
+{
+	calc_zero_flag();
+
+	if (augend > m_ALU)
+		SET(m_STATUS, C_FLAG);
+	else
+		CLR(m_STATUS, C_FLAG);
+
+	if ((augend & 0x0f) > (m_ALU & 0x0f))
+		SET(m_STATUS, DC_FLAG);
+	else
+		CLR(m_STATUS, DC_FLAG);
+}
+
+void pic16x8x_device::calc_sub_flags(u8 minuend)
+{
+	calc_zero_flag();
+
+	if (minuend < m_ALU)
+		CLR(m_STATUS, C_FLAG);
+	else
+		SET(m_STATUS, C_FLAG);
+
+	if ((minuend & 0x0f) < (m_ALU & 0x0f))
+		CLR(m_STATUS, DC_FLAG);
+	else
+		SET(m_STATUS, DC_FLAG);
+}
+
+
+
+void pic16x8x_device::set_pc(u16 addr)
+{
+	m_PCL = addr & m_program_mask;
+}
+
+u16 pic16x8x_device::pop_stack()
+{
+	m_stack_pointer--;
+	u16 data = m_STACK[m_stack_pointer];
+	return data & m_program_mask;
+}
+
+void pic16x8x_device::push_stack(u16 data)
+{
+	m_STACK[m_stack_pointer] = data & m_program_mask;
+	m_stack_pointer++;

```

Comment:
It's better to check for stack overflow here. (I haven't double-checked the datasheet for this microcontroller, but PIC17 does recognize stack overflows, though it doesn't handle them very gracefully.)

Reply:
will check the datasheet

Reply:
Done. Added some datasheet notes and add stack overflow check.